src/macro-context.js

```diff
@@ -295,16 +306,47 @@ export default class MacroContext {
 
   _rest(enf) {
     const priv = privateData.get(this);
-    if(priv.backup === enf) {
+    if(priv.markers.get(priv.startMarker) === enf) {
       return priv.enf.rest;
     }
     throw Error("Unauthorized access!");
   }
 
-  reset() {
+  reset(marker) {
+    const priv = privateData.get(this);
+    let enf;
+    if(marker == null) {
+      // go to the beginning
+      enf = priv.markers.get(priv.startMarker);
+    } else if(marker && marker instanceof Marker) {
+      // marker could be from another context
+      if(priv.markers.has(marker)) {
+        enf = priv.markers.get(marker);
+      } else {
+        throw new Error('marker must originate from this context');
+      }
+    } else {
+      throw new Error('marker must be an instance of Marker');
+    }
+    priv.enf = cloneEnforester(enf);
+  }
+
+  mark() {
     const priv = privateData.get(this);
-    const { rest, prev, context } = priv.backup;
-    priv.enf = new Enforester(rest, prev, context);
+    let marker;
+    if(priv.enf.rest.size === priv.markers.get(priv.startMarker).rest.size) {
```

Comment:
Do you mean store a reference to the first `enf`? That is what `startMarker` is the key for. I should add a comment, but the idea is that marking at the beginning shouldn't happen more than once.
The reason I can't just compare `enf` and `markers.get(startMarker)` is that they aren't the same enforester. I guess I could compare the lists and not the size.